modules/node-shared/src/main/scala/io/constellationnetwork/node/shared/infrastructure/consensus/ValidationErrorStorage.scala

```diff
@@ -0,0 +1,50 @@
+package io.constellationnetwork.node.shared.infrastructure.consensus
+
+import java.time.Instant
+
+import cats.data.NonEmptyList
+import cats.effect.kernel.{Async, Clock, Ref}
+import cats.syntax.flatMap._
+import cats.syntax.functor._
+
+import eu.timepit.refined.auto.autoUnwrap
+import eu.timepit.refined.types.all.PosInt
+
+trait ValidationErrorStorage[F[_], Artifact, ValidationError] {
+  def addValidationError(artifact: Artifact, errors: ValidationError): F[Unit]
+  def addValidationErrors(errors: NonEmptyList[(Artifact, ValidationError)]): F[Unit]
+  def allValidationErrors: F[Vector[ValidationErrorStorageEntry[Artifact, ValidationError]]]
+}
+
+case class ValidationErrorStorageEntry[Artifact, ValidationError](artifact: Artifact, validationError: ValidationError, timestamp: Instant)
+
+object ValidationErrorStorage {
+
+  def make[F[_]: Async, Artifact, ValidationError](
+    maxSize: PosInt
+  ): F[ValidationErrorStorage[F, Artifact, ValidationError]] =
+    for {
+      validationErrorsR <- Ref[F].of(Vector.empty[ValidationErrorStorageEntry[Artifact, ValidationError]])
+    } yield
+      new ValidationErrorStorage[F, Artifact, ValidationError] {
+        override def addValidationError(artifact: Artifact, error: ValidationError): F[Unit] =
+          for {
+            timestamp <- Clock[F].realTimeInstant
+            res <-
+              validationErrorsR.update { v =>
+                val bounded = if (v.size >= maxSize) v.dropRight(v.size - maxSize + 1) else v
+                ValidationErrorStorageEntry(artifact, error, timestamp) +: bounded
+              }
+          } yield res
+
+        override def addValidationErrors(errors: NonEmptyList[(Artifact, ValidationError)]): F[Unit] = {
+          val (headArtifact, headError) = errors.head
+          errors.foldLeft(addValidationError(headArtifact, headError)) {
+            case (acc, (artifact, error)) =>
+              acc >> addValidationError(artifact, error)
+          }
```

Comment:
instead of extracting the head and violating DRY rule, you can put an empty unit in the fold as a zero value and fold over all elements then. you can also use `traverse` probably, that would be easier than reducing values like that, since `F` is a monad and you can chain computations